### PR TITLE
AWS: Add SupportBulkOperation interface. S3FileIO implementation will perform a batch deletion using RemoveObjects API

### PR DESCRIPTION
**Pull request by amogh-jahagirdar**
Starting a draft PR for performing batch deletion for S3 objects. Related issue: https://github.com/apache/iceberg/issues/4012

This can be useful for the expire snapshots and remove orphan files operations. In this PR we update the FileIO interface and add a S3 implementation to perform the batch removal. Leaving it in draft until I add some unit tests for the batch removal implementation for S3, wanted to get some initial feedback on the code.

In other PRs we can tackle updating the actions for expiring snapshots and removing orphan files. My thoughts there are we do have a separate batching mechanism within the action implementation (for an action a user would specify a batch size (defaulting to 1). Tasks looks like it is generic and for the parameter we could partition the given input list into batches, and then we could pass in a function which accepts a list of strings for performing the batch deletion. This would only be done in the case the batch size is greater than 1. Any thoughts?

@szehon-ho @dramaticlly @jackye1995

### CODE REVIEW

File: aws/src/main/java/org/apache/iceberg/aws/AwsProperties.java [hunk 1]
```diff
@@ -309,6 +317,14 @@ public String s3FileIoSseKey() {
     return s3FileIoSseKey;
   }
 
+  public int s3DeletionBatchSize() {
+    return s3DeletionBatchSize;
+  }
+
+  public void setS3DeletionBatchSize(int deletionBatchSize) {
+    this.s3DeletionBatchSize = deletionBatchSize;
```
Comment (amogh-jahagirdar) [inline aws/src/main/java/org/apache/iceberg/aws/AwsProperties.java:325]: need to add validation here

File: api/src/main/java/org/apache/iceberg/io/FileIO.java [hunk 2]
```diff
@@ -68,6 +68,15 @@ default void deleteFile(OutputFile file) {
   default void initialize(Map<String, String> properties) {
   }
 
+  /**
+   * Delete the files at the given path.
+   */
+  default void deleteFiles(Iterable<String> pathsToDelete) {
```
Comment (amogh-jahagirdar) [inline api/src/main/java/org/apache/iceberg/io/FileIO.java:74]: I'm wondering if it makes more sense to just have this interface take a set of paths to delete. That way we don't have to worry about de-duping ourselves. I would imagine this is something that would be common across all FileIO.

File: aws/src/main/java/org/apache/iceberg/aws/AwsProperties.java [hunk 3]
```diff
@@ -213,13 +213,20 @@
    * Enables eTag checks for S3 PUT and MULTIPART upload requests.
    */
   public static final String S3_CHECKSUM_ENABLED = "s3.checksum-enabled";
+
+  /**
+   * Sets the size for batch deletion.
+   */
+  public static final int S3_DELETION_BATCH_SIZE_DEFAULT = 10000;
```
Comment (amogh-jahagirdar) [inline aws/src/main/java/org/apache/iceberg/aws/AwsProperties.java:220]: should be 1000 instead of 10000

Comment (dramaticlly) [reply aws/src/main/java/org/apache/iceberg/aws/AwsProperties.java:220]: nit, might worth putting a link to s3 documentation to reference where is 1000 come from https://docs.aws.amazon.com/AmazonS3/latest/API/API_DeleteObjects.html

File: aws/src/main/java/org/apache/iceberg/aws/AwsProperties.java (hunk 1 above)
Comment (dramaticlly) [reply aws/src/main/java/org/apache/iceberg/aws/AwsProperties.java:325]: curious to know, shall we allow it also to be set from aws properties  with string k/v Map?

File: api/src/main/java/org/apache/iceberg/io/FileIO.java (hunk 2 above)
Comment (dramaticlly) [reply api/src/main/java/org/apache/iceberg/io/FileIO.java:74]: I guess it's fine to leave it as Iterable as I feel the java set also implements the Iterable interface, so if needed the consumer of this method can pass the set for de-duplication purpose. If there's a need to preserve the ordering, then they can pass other iterable like list as well. From correctness perspective, the concrete file system shall be handle the duplicate file path for deletion.

File: aws/src/main/java/org/apache/iceberg/aws/s3/S3FileIO.java [hunk 4]
```diff
@@ -94,6 +106,53 @@ public void deleteFile(String path) {
     client().deleteObject(deleteRequest);
   }
 
+  /**
+   * Performs a batch deletion of the given paths. This implementation does not require
+   * that the passed in paths refer to objects which are in the same bucket.
+   * @param paths poths to delete
```
Comment (dramaticlly) [inline aws/src/main/java/org/apache/iceberg/aws/s3/S3FileIO.java:112]: nit: I would probably worded a bit differently like `Performs a batch deletion of given paths grouped by the bucket. Deletion will trigger whenever configured batch size reached for given bucket or for each bucket at the end.

also probably typo in param comment? `s/poths/paths`

**Review by dramaticlly — COMMENTED** [review]
Thank you @amogh-jahagirdar , left some nitpicky comments

File: api/src/main/java/org/apache/iceberg/io/FileIO.java (hunk 2 above)
Comment (jackye1995) [reply api/src/main/java/org/apache/iceberg/io/FileIO.java:74]: Would S3 API fail us if there is any duplicated path in the input request? Have you tried out?

File: aws/src/main/java/org/apache/iceberg/aws/AwsProperties.java [hunk 5]
```diff
@@ -213,13 +213,20 @@
    * Enables eTag checks for S3 PUT and MULTIPART upload requests.
    */
   public static final String S3_CHECKSUM_ENABLED = "s3.checksum-enabled";
+
+  /**
+   * Sets the size for batch deletion.
+   */
```
Comment (jackye1995) [inline aws/src/main/java/org/apache/iceberg/aws/AwsProperties.java:219]: I think we also need a config for this

File: aws/src/main/java/org/apache/iceberg/aws/AwsProperties.java [hunk 6]
```diff
@@ -213,13 +213,20 @@
...
+  public static final int S3_DELETION_BATCH_SIZE_DEFAULT = 10000;
+
   public static final boolean CLIENT_ENABLE_ETAG_CHECK_DEFAULT = false;
 
   private String s3FileIoSseType;
   private String s3FileIoSseKey;
   private String s3FileIoSseMd5;
   private int s3FileIoMultipartUploadThreads;
   private int s3FileIoMultiPartSize;
+  private int s3DeletionBatchSize;
```
Comment (jackye1995) [inline aws/src/main/java/org/apache/iceberg/aws/AwsProperties.java:229]: `s3FileIoDeleteBatchSize`?

**Review by jackye1995 — COMMENTED** [review]
Thanks for working on this! I think we need some integration tests especially for:
(1) verify the limit of S3 batch delete, if 1000 batch size is good for common usage
(2) do we need to dedupe on client side for the paths input. Maybe we can just ignore, and then catch object not found error that is more efficient.

File: aws/src/main/java/org/apache/iceberg/aws/s3/S3FileIO.java [hunk 7]
```diff
@@ -94,6 +106,53 @@ public void deleteFile(String path) {
...
+      }
+      bucketToObjects.get(bucket).add(objectKey);
+    }
+    // Delete the remainder
+    bucketToObjects
+        .asMap()
+        .forEach((bucket, objectsToDelete) -> deleteObjectsInBucket(bucket, objectsToDelete));
+  }
+
+  private void deleteObjectsInBucket(String bucket, Collection<String> objects) {
```
[the remark this answers is not in the document]
Comment (amogh-jahagirdar) [reply aws/src/main/java/org/apache/iceberg/aws/s3/S3FileIO.java:134]: We could, the exception message should then encapsulate the failures that happened across all the buckets which had some failures.

File: aws/src/main/java/org/apache/iceberg/aws/s3/S3FileIO.java (hunk 4 above)
Comment (amogh-jahagirdar) [reply aws/src/main/java/org/apache/iceberg/aws/s3/S3FileIO.java:112]: Good catch, and yeah I think the comment you have is more clear.

File: aws/src/main/java/org/apache/iceberg/aws/AwsProperties.java (hunk 3 above)
Comment (amogh-jahagirdar) [reply aws/src/main/java/org/apache/iceberg/aws/AwsProperties.java:220]: Very true, I actually should not assume that it's a good default until I do some benchmarking.

File: aws/src/main/java/org/apache/iceberg/aws/AwsProperties.java (hunk 5 above)
Comment (amogh-jahagirdar) [reply aws/src/main/java/org/apache/iceberg/aws/AwsProperties.java:231]: yes definitley, added!

**Comment by amogh-jahagirdar** [comment]
Sorry for the delay folks. Few updates:

1.)  updated the PR with some integration tests and more unit tests. 

2.) The deletion batch size is configurable through s3.delete.batch-size.

3.) The default is 250 instead of 1000. Tbh I think some rigorous benchmarking should be done here. I set to 250 mostly mimicking the similar change done in hadoop-aws https://github.com/apache/hadoop/commit/56dee667707926f3796c7757be1a133a362f05c9 which also used to perform batch deletions in 1000 until encountering major throttling issues. For reference if there are N keys in a batch, this will uses N requests in your throughput calculation done by S3 for controlling throttling. S3 limitations are 3500 TPS per prefix. So if we did 1000, in the worst case where most of the keys fall in the same prefix (if somebody has a hive-like file structure) then we would easily hit this limitation easily. If prefixes are better distributed we could get more throughput, but don't think we should rely on this assumption.

File: aws/src/main/java/org/apache/iceberg/aws/AwsProperties.java (hunk 1 above)
Comment (amogh-jahagirdar) [reply aws/src/main/java/org/apache/iceberg/aws/AwsProperties.java]: Yes it definitley should. In my latest revision I have added a property!

File: api/src/main/java/org/apache/iceberg/io/FileIO.java (hunk 2 above)
Comment (amogh-jahagirdar) [reply api/src/main/java/org/apache/iceberg/io/FileIO.java:74]: For both duplicated paths and in the case the key does not exist (already deleted) S3 will not fail.

File: aws/src/integration/java/org/apache/iceberg/aws/s3/TestS3FileIOIntegration.java [hunk 8]
```diff
@@ -208,8 +212,49 @@ public void testClientFactorySerialization() throws Exception {
     validateRead(fileIO2);
   }
 
+  @Test
+  public void testDeleteFilesMultipleBatches() throws Exception {
+    S3FileIO s3FileIO = new S3FileIO(clientFactory::s3, getDeletionTestProperties());
+    testDeleteFiles(6, s3FileIO);
```
Comment (dramaticlly) [inline aws/src/integration/java/org/apache/iceberg/aws/s3/TestS3FileIOIntegration.java:218]: nit, maybe I would think about put it like `deletionBatchSize * 2`

same as below 
- for lessThan, we can do `deletionBatchSize-1`
- for SingleWithRemainer do `deletionBatchSize+1`

so it shall help with understanding and hopefully more robust once `deletionBatchSize` get changed

File: aws/src/main/java/org/apache/iceberg/aws/AwsProperties.java (hunk 5 above)
Comment (dramaticlly) [reply aws/src/main/java/org/apache/iceberg/aws/AwsProperties.java:231]: Nice finding, I would even consider put a comment to reference where is 250 comes from https://github.com/apache/hadoop/commit/56dee667707926f3796c7757be1a133a362f05c9 until benchmark is completed

**Review by dramaticlly — APPROVED** [review]
LGTM, thank for the changes

File: aws/src/integration/java/org/apache/iceberg/aws/s3/TestS3FileIOIntegration.java (hunk 8 above)
Comment (amogh-jahagirdar) [reply aws/src/integration/java/org/apache/iceberg/aws/s3/TestS3FileIOIntegration.java:218]: Yeah definitley, this is much better

File: aws/src/main/java/org/apache/iceberg/aws/AwsProperties.java [hunk 9]
```diff
@@ -300,6 +312,11 @@ public AwsProperties(Map<String, String> properties) {
     this.isS3ChecksumEnabled = PropertyUtil.propertyAsBoolean(properties, S3_CHECKSUM_ENABLED,
         S3_CHECKSUM_ENABLED_DEFAULT);
 
+    this.s3FileIoDeleteBatchSize = PropertyUtil.propertyAsInt(properties, S3FILEIO_DELETE_BATCH_SIZE,
+        S3FILEIO_DELETE_BATCH_SIZE_DEFAULT);
+    Preconditions.checkArgument(s3FileIoDeleteBatchSize > 0 && s3FileIoDeleteBatchSize < 1000,
```
Comment (jackye1995) [inline aws/src/main/java/org/apache/iceberg/aws/AwsProperties.java:317]: 1000 should be a static variable

File: aws/src/main/java/org/apache/iceberg/aws/s3/S3FileIO.java [hunk 10]
```diff
@@ -100,6 +114,66 @@ public void deleteFile(String path) {
...
+    List<List<String>> remainderFailedObjects = bucketToObjects
+        .asMap()
+        .entrySet()
+        .stream()
+        .map(entry -> deleteObjectsInBucket(entry.getKey(), entry.getValue()))
+        .collect(Collectors.toList());
+
+    remainderFailedObjects.forEach(failedDeletions::addAll);
+    if (!failedDeletions.isEmpty()) {
+      throw new S3BatchDeletionException(String.format("Failed to delete %d objects. Failed objects: %s",
```
Comment (jackye1995) [inline aws/src/main/java/org/apache/iceberg/aws/s3/S3FileIO.java:151]: I think this will be not just for S3, we might want to add it to the api package as a generic `BatchDeleteFailedException`. We can also make constructor fit the input because this is a custom exception class. 

Printing all the failed object paths might be a bit too much, if there are many file paths it's going to make log explode, I think a count is sufficient.

File: aws/src/main/java/org/apache/iceberg/aws/s3/S3FileIO.java [hunk 11]
```diff
@@ -100,6 +114,66 @@ public void deleteFile(String path) {
...
+          .delete(Delete.builder().objects(objectIds).build())
+          .build();
+      DeleteObjectsResponse response = client().deleteObjects(deleteObjectsRequest);
+      if (response.hasErrors()) {
+        return response.errors()
+            .stream()
+            .map(S3Error::key)
+            .collect(Collectors.toList());
+      }
+    }
```
Comment (jackye1995) [inline aws/src/main/java/org/apache/iceberg/aws/s3/S3FileIO.java:173]: nit: newline after `if` block

File: aws/src/main/java/org/apache/iceberg/aws/s3/S3FileIO.java [hunk 12]
```diff
@@ -100,6 +114,66 @@ public void deleteFile(String path) {
...
+      String bucket = location.bucket();
+      String objectKey = location.key();
+      Set<String> objectsInBucket = bucketToObjects.get(bucket);
+      if (objectsInBucket.size() == awsProperties.s3FileIoDeleteBatchSize()) {
+        List<String> failedDeletionsForBatch = deleteObjectsInBucket(bucket, objectsInBucket);
+        failedDeletions.addAll(failedDeletionsForBatch);
+        bucketToObjects.removeAll(bucket);
+      }
+      bucketToObjects.get(bucket).add(objectKey);
+    }
```
Comment (jackye1995) [inline aws/src/main/java/org/apache/iceberg/aws/s3/S3FileIO.java:140]: nit: newline after `for` block

**Review by jackye1995 — COMMENTED** [review]
overall this looks good to me!

File: aws/src/main/java/org/apache/iceberg/aws/s3/S3FileIO.java (hunk 10 above)
Comment (amogh-jahagirdar) [reply aws/src/main/java/org/apache/iceberg/aws/s3/S3FileIO.java:151]: Agree on generic exception, but I think we probably want a different name because it's not necessary that deleteFiles(files) corresponds to batch deletion for other fileIO implementations. the default implementation is to iteratively delete files one by one. So we could define a DeletionFailedException?

**Comment by amogh-jahagirdar** [comment]
Updated. For some reason I had the "core" label on the commit message, but this is API and AWS module updates. Updated the commit message to reflect this.

File: aws/src/main/java/org/apache/iceberg/aws/AwsProperties.java [hunk 13]
```diff
@@ -300,6 +321,12 @@ public AwsProperties(Map<String, String> properties) {
     this.isS3ChecksumEnabled = PropertyUtil.propertyAsBoolean(properties, S3_CHECKSUM_ENABLED,
         S3_CHECKSUM_ENABLED_DEFAULT);
 
+    this.s3FileIoDeleteBatchSize = PropertyUtil.propertyAsInt(properties, S3FILEIO_DELETE_BATCH_SIZE,
+        S3FILEIO_DELETE_BATCH_SIZE_DEFAULT);
+    Preconditions.checkArgument(s3FileIoDeleteBatchSize > 0 &&
+        s3FileIoDeleteBatchSize <= S3FILEIO_DELETE_MAX_BATCH_SIZE,
```
Comment (amogh-jahagirdar) [inline aws/src/main/java/org/apache/iceberg/aws/AwsProperties.java:327]: Unfortunately looks like the only decent line break choice for this (both conditions won't fit on the same line).

File: aws/src/main/java/org/apache/iceberg/aws/AwsProperties.java [hunk 14]
```diff
@@ -171,6 +171,25 @@
...
+   * Refer to https://github.com/apache/hadoop/commit/56dee667707926f3796c7757be1a133a362f05c9
+   * for more details on why this value was chosen.
+   */
+  public static final int S3FILEIO_DELETE_BATCH_SIZE_DEFAULT = 250;
+
+  /**
+   * Max possible batch size for deletion. Currently, a max of 1000 keys can be deleted in one batch.
+   * https://docs.aws.amazon.com/AmazonS3/latest/API/API_DeleteObjects.html
+   */
+  public static final int S3FILEIO_DELETE_MAX_BATCH_SIZE = 1000;
```
Comment (jackye1995) [inline aws/src/main/java/org/apache/iceberg/aws/AwsProperties.java:191]: nit `S3FILEIO_DELETE_BATCH_SIZE_MAX`

File: aws/src/test/java/org/apache/iceberg/aws/TestAwsProperties.java [hunk 15]
```diff
@@ -97,4 +97,24 @@ public void testS3MultipartThresholdFactorLessThanOne() {
         () -> new AwsProperties(map));
   }
 
+  @Test
+  public void testS3FileIoDeleteBatchSizeTooLarge() {
+    Map<String, String> map = Maps.newHashMap();
+    map.put(AwsProperties.S3FILEIO_DELETE_BATCH_SIZE, "2000");
+    AssertHelpers.assertThrows("should not accept batch size greater than 1000",
+            IllegalArgumentException.class,
```
Comment (jackye1995) [inline aws/src/test/java/org/apache/iceberg/aws/TestAwsProperties.java:105]: the indentations in tests seem to be off? I think it should have 4 indent

File: api/src/main/java/org/apache/iceberg/io/FileIO.java [hunk 16]
```diff
@@ -68,6 +68,15 @@ default void deleteFile(OutputFile file) {
   default void initialize(Map<String, String> properties) {
   }
 
+  /**
+   * Delete the files at the given path.
```
Comment (jackye1995) [inline api/src/main/java/org/apache/iceberg/io/FileIO.java:72]: nit: can add `@throws FileIODeletionException`

**Review by jackye1995 — APPROVED** [review]
Mostly looks good to me, just a few nitpicking. Thanks for working for this!

File: aws/src/main/java/org/apache/iceberg/aws/AwsProperties.java (hunk 14 above)
Comment (amogh-jahagirdar) [reply aws/src/main/java/org/apache/iceberg/aws/AwsProperties.java]: Updated, looks like this is the established naming convention for max,min. values.

File: aws/src/test/java/org/apache/iceberg/aws/TestAwsProperties.java (hunk 15 above)
Comment (amogh-jahagirdar) [reply aws/src/test/java/org/apache/iceberg/aws/TestAwsProperties.java]: Updated. Strange, I think when I was working on adding these tests, my IDE was using different the default indentation settings rather than the Iceberg ones.

File: api/src/main/java/org/apache/iceberg/io/FileIO.java (hunk 16 above)
Comment (amogh-jahagirdar) [reply api/src/main/java/org/apache/iceberg/io/FileIO.java:73]: Updated

File: api/src/main/java/org/apache/iceberg/exceptions/FileIODeletionException.java [hunk 17]
```diff
@@ -0,0 +1,28 @@
...
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.exceptions;
+
+import java.util.Collection;
+
+public class FileIODeletionException extends RuntimeException {
+  public FileIODeletionException(Collection<String> failedDeletions) {
```
Comment (rdblue) [inline api/src/main/java/org/apache/iceberg/exceptions/FileIODeletionException.java:25]: How about `BulkDeleteFailureException`? I don't see a need to scope this to `FileIO`.

File: api/src/main/java/org/apache/iceberg/io/FileIO.java [hunk 18]
```diff
@@ -68,6 +69,17 @@ default void deleteFile(OutputFile file) {
...
+   *
+   * @throws FileIODeletionException If failures are encountered when deleting the given files
```
Comment (rdblue) [inline api/src/main/java/org/apache/iceberg/io/FileIO.java:75]: This needs to document `pathsToDelete` as well.

File: aws/src/main/java/org/apache/iceberg/aws/s3/S3FileIO.java [hunk 19]
```diff
@@ -100,6 +115,67 @@ public void deleteFile(String path) {
...
+
+    // Delete the remainder
+    List<List<String>> remainderFailedObjects = bucketToObjects
+        .asMap()
+        .entrySet()
+        .stream()
+        .map(entry -> deleteObjectsInBucket(entry.getKey(), entry.getValue()))
```
Comment (rdblue) [inline aws/src/main/java/org/apache/iceberg/aws/s3/S3FileIO.java:148]: Is there a `flatMap` method that could avoid running `forEach` after this?

File: aws/src/main/java/org/apache/iceberg/aws/s3/S3FileIO.java [hunk 20]
```diff
@@ -100,6 +115,67 @@ public void deleteFile(String path) {
...
+        bucketToObjects.removeAll(bucket);
+      }
+      bucketToObjects.get(bucket).add(objectKey);
+    }
+
+    // Delete the remainder
+    List<List<String>> remainderFailedObjects = bucketToObjects
+        .asMap()
+        .entrySet()
+        .stream()
```
Comment (rdblue) [inline aws/src/main/java/org/apache/iceberg/aws/s3/S3FileIO.java:147]: I'd probably move these 3 calls to the previous line since they aren't really interesting on their own.

File: aws/src/main/java/org/apache/iceberg/aws/s3/S3FileIO.java [hunk 21]
```diff
@@ -100,6 +115,67 @@ public void deleteFile(String path) {
...
+        .map(entry -> deleteObjectsInBucket(entry.getKey(), entry.getValue()))
+        .collect(Collectors.toList());
+
+    remainderFailedObjects.forEach(failedDeletions::addAll);
+    if (!failedDeletions.isEmpty()) {
+      throw new FileIODeletionException(failedDeletions);
+    }
+
```
Comment (rdblue) [inline aws/src/main/java/org/apache/iceberg/aws/s3/S3FileIO.java:155]: Nit: unnecessary newline.

File: aws/src/main/java/org/apache/iceberg/aws/s3/S3FileIO.java [hunk 22]
```diff
@@ -100,6 +115,67 @@ public void deleteFile(String path) {
...
+  public void deleteFiles(Iterable<String> paths) {
+    SetMultimap<String, String> bucketToObjects = Multimaps.newSetMultimap(Maps.newHashMap(), Sets::newHashSet);
+    List<String> failedDeletions = Lists.newArrayList();
+    for (String path : paths) {
+      S3URI location = new S3URI(path);
+      String bucket = location.bucket();
+      String objectKey = location.key();
+      Set<String> objectsInBucket = bucketToObjects.get(bucket);
+      if (objectsInBucket.size() == awsProperties.s3FileIoDeleteBatchSize()) {
+        List<String> failedDeletionsForBatch = deleteObjectsInBucket(bucket, objectsInBucket);
```
Comment (rdblue) [inline aws/src/main/java/org/apache/iceberg/aws/s3/S3FileIO.java:136]: Normally, we will retry deletes unless there is some exception that prevents it (like FileNotFound or Unauthorized). Why not do that here? Is that a follow-up issue?

File: aws/src/main/java/org/apache/iceberg/aws/s3/S3FileIO.java [hunk 23]
```diff
@@ -100,6 +115,67 @@ public void deleteFile(String path) {
...
+  /**
+   * Deletes the given paths in a batched manner.
+   * <p>
+   * The paths are grouped by bucket, and deletion is triggered when we either reach the configured batch size
+   * or have a final remainder batch for each bucket.
+   *
+   * @param paths paths to delete
+   */
+  @Override
+  public void deleteFiles(Iterable<String> paths) {
```
Comment (rdblue) [inline aws/src/main/java/org/apache/iceberg/aws/s3/S3FileIO.java:127]: If the batch size is 1, I think this should fall back to the normal delete implementation.

File: aws/src/main/java/org/apache/iceberg/aws/s3/S3FileIO.java [hunk 24]
```diff
@@ -100,6 +115,67 @@ public void deleteFile(String path) {
...
+          .build();
+      DeleteObjectsResponse response = client().deleteObjects(deleteObjectsRequest);
+      if (response.hasErrors()) {
+        return response.errors()
+            .stream()
+            .map(S3Error::key)
+            .collect(Collectors.toList());
+      }
+    }
+    return Lists.newArrayList();
```
Comment (rdblue) [inline aws/src/main/java/org/apache/iceberg/aws/s3/S3FileIO.java:176]: Nit: missing newline between control flow block and the return statement.

File: aws/src/main/java/org/apache/iceberg/aws/s3/S3FileIO.java [hunk 25]
```diff
@@ -100,6 +115,67 @@ public void deleteFile(String path) {
...
+          .collect(Collectors.toList());
+      DeleteObjectsRequest deleteObjectsRequest = DeleteObjectsRequest.builder()
+          .bucket(bucket)
+          .delete(Delete.builder().objects(objectIds).build())
+          .build();
+      DeleteObjectsResponse response = client().deleteObjects(deleteObjectsRequest);
+      if (response.hasErrors()) {
+        return response.errors()
+            .stream()
+            .map(S3Error::key)
```
Comment (rdblue) [inline aws/src/main/java/org/apache/iceberg/aws/s3/S3FileIO.java:172]: Since this is just the key, the exception isn't going to show the bucket that these keys are in. Can you handle this and reconstruct the full S3 location?

Comment (amogh-jahagirdar) [reply aws/src/main/java/org/apache/iceberg/aws/s3/S3FileIO.java:172]: Ah yeah that's true. We should construct the full S3 path

File: aws/src/main/java/org/apache/iceberg/aws/s3/S3FileIO.java (hunk 23 above)
Comment (amogh-jahagirdar) [reply aws/src/main/java/org/apache/iceberg/aws/s3/S3FileIO.java:127]: We can. The current logic is generic but for the case where batch size is 1 there's more unnecessary state management, so I'm in favor of falling back to just looping and calling the existing delete.

Comment (amogh-jahagirdar) [reply aws/src/main/java/org/apache/iceberg/aws/s3/S3FileIO.java:127]: Actually, one issue with falling back to the normal delete implementation in case the batch size is 1 is we currently don't have any expectation on the exception being surfaced by FileIo#deleteFile in case of a failure during deletion.
 
With deleteFiles we are keeping track of failures and then surfacing them at the end, so in the case the batch size is 1 we would have to catch generic Exceptions and keep track of failures that way which seems more messy. So now I'm thinking it makes more sense to keep the generic batch removal logic (the multi-map state management is light-weight). Let me know your thoughts @rdblue

File: api/src/main/java/org/apache/iceberg/exceptions/FileIODeletionException.java (hunk 17 above)
Comment (amogh-jahagirdar) [reply api/src/main/java/org/apache/iceberg/exceptions/FileIODeletionException.java:25]: Originally I was thinking it makes more sense to be more generic because deleteFiles does not impose that the underlying implementation performs bulk removals against the underlying storage. But on second thought, I do agree that scoping something to FileIO for the first implementation of this API may be unnecessarily broad. Will update to use BulkDeleteFailureException.

File: aws/src/main/java/org/apache/iceberg/aws/s3/S3FileIO.java (hunk 22 above)
Comment (amogh-jahagirdar) [reply aws/src/main/java/org/apache/iceberg/aws/s3/S3FileIO.java:136]: I was thinking it would be up to the provider of the S3 client who would configure the retry policy on the client. Is that something within the scope of FileIO? If so I think that's something we could tackle in a follow-on.

Someone could use a custom AwsClientFactory . The DefaultAwsClientFactory will create an S3 client with the default retry policy which would retry on the failures mentioned in https://docs.aws.amazon.com/AWSJavaSDK/latest/javadoc/com/amazonaws/retry/PredefinedRetryPolicies.SDKDefaultRetryCondition.html.


So basically 5xx errors like service unavailable, throttling, clock-skew etc would be retried. Failures such as bucket not existing, or unauthorized 4xx errors would not be retried by default. @jackye1995 @rdblue thoughts?

File: aws/src/main/java/org/apache/iceberg/aws/s3/S3FileIO.java (hunk 20 above)
Comment (amogh-jahagirdar) [reply aws/src/main/java/org/apache/iceberg/aws/s3/S3FileIO.java:147]: Updated

File: aws/src/main/java/org/apache/iceberg/aws/s3/S3FileIO.java (hunk 19 above)
Comment (amogh-jahagirdar) [reply aws/src/main/java/org/apache/iceberg/aws/s3/S3FileIO.java:148]: I wasn't able to find a goodway with flatmap, but there is a more succinct way if we do .asMap().forEach(). Let me know what you think

File: aws/src/main/java/org/apache/iceberg/aws/s3/S3FileIO.java (hunk 24 above)
Comment (amogh-jahagirdar) [reply aws/src/main/java/org/apache/iceberg/aws/s3/S3FileIO.java:176]: Updated

File: api/src/main/java/org/apache/iceberg/io/FileIO.java (hunk 18 above)
Comment (amogh-jahagirdar) [reply api/src/main/java/org/apache/iceberg/io/FileIO.java]: Updated

File: aws/src/main/java/org/apache/iceberg/aws/s3/S3FileIO.java (hunk 22 above)
Comment (amogh-jahagirdar) [reply aws/src/main/java/org/apache/iceberg/aws/s3/S3FileIO.java:135]: Actually, I realized the current default implementation would be inconsistent with what I just mentioned. for default, we just loop over the files and delete, we don't surface the failure at the end; if there's a failure it will be surfaced immediately. I'm still leaning towards the deleteFiles semantic being a best effort deletion attempt on all files in the list (surfacing failures at the end). So I am more leaning towards changing the default implementation. 

Let me know your thoughts if you agree on this semantic for deleteFiles @rdblue @jackye1995 @rdblue

Comment (danielcweeks) [reply aws/src/main/java/org/apache/iceberg/aws/s3/S3FileIO.java:135]: I feel like we should not rely on failure/exceptions and keep delete "best effort".  The reason for this is that the cleanup being done using FileIO delete (whether it's for failed/speculative tasks or other maintenance type clean) up is not a requirement for iceberg to function and will lead to spurious errors and inconsistency (as we see the normal delete path doesn't report errors in multiple implementations).

Even in a delete failure case, the required action is to run the maintenance procedures to perform orphan file cleanup, so we end up propagating errors that could result in job failure due to issues that would normally not affect a job.

I think it's fine to log failed deletes for debugging purposes, but we shouldn't rely on the deletes.

Alternatively, we could keep this bulk functionality out of the FileIO implementation and add a "feature" interface like `SupportsBulkDelete` that can be used for S3 specifically.  Not all implementations of FileIO have a concept of bulk delete (S3 is the only one I'm aware of).  That would allow us to have a stricter contract around deletes for cleanup procedures.

I'd actually prefer the latter as it keeps FileIO cleaner and we can separate contract behaviors without introducing inconsistency.  Actions may need additional feature not available (e.g. listing) that we don't want to add to the FileIO contract directly.

**Comment by amogh-jahagirdar** [comment]
@rdblue @danielcweeks @jackye1995 Sorry for the delay, I've updated the PR. Some fundamental changes from before:

1.) No deleteFiles on FileIO. 

2.) public BulkOperation interface which is currently only implemented for S3FileIO. batch deletion exception are only for the S3 case (package private). 

It does make it a bit cumbersome when we want to do batch deletion in removing orphan files, because we would have to do instanceOf check.

File: api/src/main/java/org/apache/iceberg/aws/s3/BatchDeleteFailureException.java [hunk 26]
```diff
@@ -0,0 +1,35 @@
...
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.aws.s3;
+
+import java.util.Collection;
+
+class BatchDeleteFailureException extends RuntimeException {
```
Comment (amogh-jahagirdar) [inline api/src/main/java/org/apache/iceberg/aws/s3/BatchDeleteFailureException.java:24]: I think this should be public instead of package private. Otherwise an external system would not be able to leverage the numberFailedObjects?

File: aws/src/main/java/org/apache/iceberg/aws/s3/S3FileIO.java (hunk 22 above)
Comment (amogh-jahagirdar) [reply aws/src/main/java/org/apache/iceberg/aws/s3/S3FileIO.java:135]: @danielcweeks @jackye1995 @rdblue Updated the latest so that it's contained within a separate feature interface, let me know what you think!

Comment (danielcweeks) [reply aws/src/main/java/org/apache/iceberg/aws/s3/S3FileIO.java:135]: It looks like we're collecting all of the failed deletions in `failedDeletions` but not really doing anything with that.  Since the interface is intended to be iterable and could potentially have a lot of delete values, I would suggest just warn logging the path and keeping a count as opposed to keeping the full list of all the paths.

Other than that, this looks good.

Comment (amogh-jahagirdar) [reply aws/src/main/java/org/apache/iceberg/aws/s3/S3FileIO.java:138]: Makes sense. Updated to logging the paths which we failed to delete, and keeping just a count of the failures.

Comment (amogh-jahagirdar) [reply aws/src/main/java/org/apache/iceberg/aws/s3/S3FileIO.java:138]: @danielcweeks It looks like GCS does support batch operations. https://cloud.google.com/storage/docs/batch.

File: api/src/main/java/org/apache/iceberg/aws/s3/BatchDeleteFailureException.java [hunk 27]
```diff
@@ -0,0 +1,33 @@
...
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.aws.s3;
+
+public class BatchDeleteFailureException extends RuntimeException {
```
Comment (jackye1995) [inline api/src/main/java/org/apache/iceberg/aws/s3/BatchDeleteFailureException.java:22]: since now we have `SupportsBulkOperations`, should we also move this to core API as `BulkDeleteFailureException`, and specify `void deleteFiles(Iterable<String> pathsToDelete) throws BulkDeleteFailureException`?

**Review by danielcweeks — APPROVED** [review]
